File: src/utils/productRecommendations.ts
import { normalizeCategoryId } from "../data/categories";
import type { Product } from "../types/product";
import { getBrowsingHistory } from "./userStorage";

const DEFAULT_LIMIT = 28;

function popularityScore(p: Product): number {
  return p.reviewCount * p.rating + (p.prime ? 50 : 0);
}

/**
 * Cross-category picks: recent browsing history first, then popular items outside the current aisle.
 * Excludes the current product and IDs already shown in "Related products".
 */
export function pickCustomersAlsoViewed(
  product: Product,
  allProducts: Product[],
  userId: string | null | undefined,
  excludeIds: string[] = [],
  limit = DEFAULT_LIMIT
): Product[] {
  const exclude = new Set<string>([product.id, ...excludeIds]);
  const currentCategory = normalizeCategoryId(product.category);
  const picked: Product[] = [];
  const pickedIds = new Set<string>();

  const add = (p: Product | undefined) => {
    if (!p || exclude.has(p.id) || pickedIds.has(p.id)) return;
    picked.push(p);
    pickedIds.add(p.id);
  };

  for (const entry of getBrowsingHistory(userId)) {
    if (picked.length >= limit) break;
    add(allProducts.find((p) => p.id === entry.productId));
  }

  if (picked.length < limit) {
    const fillers = allProducts
      .filter(
        (p) =>
          !exclude.has(p.id) &&
          !pickedIds.has(p.id) &&
          normalizeCategoryId(p.category) !== currentCategory
      )
      .sort((a, b) => popularityScore(b) - popularityScore(a));

    for (const p of fillers) {
      if (picked.length >= limit) break;
      add(p);
    }
  }

  return picked;
}

File: mobile/lib/utils/product_recommendations.dart
import '../models/product.dart';
import '../data/kitchen_ware.dart';
import 'categories.dart';
import 'kitchen_mode.dart';

const pdpRecommendationCount = 16;

double _popularityScore(Product product) {
  return product.reviewCount * product.rating + (product.prime ? 50 : 0);
}

List<Product> _scopedCatalog(Product product, List<Product> allProducts) {
  return isKitchenProduct(product)
      ? filterKitchenProducts(allProducts)
      : excludeKitchenProducts(allProducts);
}

List<Product> pickRelatedProducts(
  Product product,
  List<Product> allProducts, {
  int limit = pdpRecommendationCount,
}) {
  final catalog = _scopedCatalog(product, allProducts);
  final related = catalog
      .where(
        (p) =>
            p.id != product.id &&
            p.inStock &&
            (isKitchenProduct(product)
                ? aisleIdFromProductId(p.id) == aisleIdFromProductId(product.id) ||
                    productMatchesCategory(p.category, product.category)
                : productMatchesCategory(p.category, product.category)),
      )
      .toList()
    ..sort((a, b) => _popularityScore(b).compareTo(_popularityScore(a)));

  if (related.length >= limit) return related.take(limit).toList();

  final exclude = {product.id, ...related.map((p) => p.id)};
  final fillers = catalog.where((p) => !exclude.contains(p.id) && p.inStock).toList()
    ..sort((a, b) => _popularityScore(b).compareTo(_popularityScore(a)));
  return [...related, ...fillers].take(limit).toList();
}

List<Product> pickYouMightAlsoLike(
  Product product,
  List<Product> allProducts, {
  Iterable<String> excludeIds = const [],
  int limit = pdpRecommendationCount,
}) {
  final catalog = _scopedCatalog(product, allProducts);
  final exclude = {product.id, ...excludeIds};
  final currentCategory = normalizeCategoryId(product.category);
  final currentAisle = aisleIdFromProductId(product.id);

  final picks = catalog
      .where((p) {
        if (exclude.contains(p.id) || !p.inStock) return false;
        if (isKitchenProduct(product)) {
          final aisle = aisleIdFromProductId(p.id);
          return aisle != null && aisle != currentAisle;
        }
        return normalizeCategoryId(p.category) != currentCategory;
      })
      .toList()
    ..sort((a, b) => _popularityScore(b).compareTo(_popularityScore(a)));

  return picks.take(limit).toList();
}
